File: src/chessie/game/clock.py
"""Chess clock with Fischer increment support."""

from __future__ import annotations

import time
from dataclasses import dataclass

from chessie.core.enums import Color
from chessie.game.interfaces import IClock, TimeControl


@dataclass(frozen=True, slots=True)
class ClockSnapshot:
    """Serializable clock state used to restore time after undo."""

    white_remaining: float
    black_remaining: float
    active_color: Color | None
    is_running: bool


class Clock(IClock):
    """Dual chess clock tracking remaining time for both players.

    Uses monotonic time for accuracy. Supports Fischer increment.
    """

    __slots__ = (
        "_time_control",
        "_remaining",
        "_active_color",
        "_last_tick",
        "_running",
    )

    def __init__(self, time_control: TimeControl) -> None:
        self._time_control = time_control
        self._remaining: dict[Color, float] = {
            Color.WHITE: time_control.initial_seconds,
            Color.BLACK: time_control.initial_seconds,
        }
        self._active_color: Color | None = None
        self._last_tick: float = 0.0
        self._running: bool = False

    # ── IClock implementation ────────────────────────────────────────────

    def start(self, color: Color) -> None:
        self._active_color = color
        self._last_tick = time.monotonic()
        self._running = True

    def stop(self) -> None:
        if self._running:
            self._consume_elapsed()
            self._running = False

    def switch(self) -> None:
        """Stop current player's clock, start the other player's."""
        if self._active_color is None:
            return
        if self._running:
            self._consume_elapsed()
        self._active_color = self._active_color.opposite
        self._last_tick = time.monotonic()

    def remaining(self, color: Color) -> float:
        if self._running and self._active_color == color:
            elapsed = time.monotonic() - self._last_tick
            return max(0.0, self._remaining[color] - elapsed)
        return max(0.0, self._remaining[color])

    def is_flag_fallen(self, color: Color) -> bool:
        return self.remaining(color) <= 0.0

    def add_increment(self, color: Color) -> None:
        self._remaining[color] += self._time_control.increment_seconds

    # ── Extra helpers ────────────────────────────────────────────────────

    @property
    def is_unlimited(self) -> bool:
        return self._time_control.initial_seconds == float("inf")

    @property
    def is_running(self) -> bool:
        return self._running

    @property
    def active_color(self) -> Color | None:
        return self._active_color

    def set_remaining(self, color: Color, seconds: float) -> None:
        """Manually override remaining time (for testing / UI override)."""
        self._remaining[color] = seconds

    def snapshot(self) -> ClockSnapshot:
        """Capture current clock state (including active side and running flag)."""
        return ClockSnapshot(
            white_remaining=self.remaining(Color.WHITE),
            black_remaining=self.remaining(Color.BLACK),
            active_color=self._active_color,
            is_running=self._running,
        )

    def restore(self, snapshot: ClockSnapshot) -> None:
        """Restore clock state previously captured with :meth:`snapshot`."""
        self._remaining[Color.WHITE] = snapshot.white_remaining
        self._remaining[Color.BLACK] = snapshot.black_remaining
        self._active_color = snapshot.active_color
        self._running = snapshot.is_running and snapshot.active_color is not None
        self._last_tick = time.monotonic()

    # ── Internal ─────────────────────────────────────────────────────────

    def _consume_elapsed(self) -> None:
        if self._active_color is None:
            return
        now = time.monotonic()
        elapsed = now - self._last_tick
        self._remaining[self._active_color] = max(
            0.0, self._remaining[self._active_color] - elapsed
        )
        self._last_tick = now

File: src/game/clock.cpp
#include <chessie/game/clock.hpp>

#include <algorithm>

namespace chessie {

Clock::Clock(TimeControl time_control) : time_control_(time_control) {
    remaining_[color_index(Color::White)] = time_control.initial_seconds;
    remaining_[color_index(Color::Black)] = time_control.initial_seconds;
}

void Clock::start(Color color) {
    active_color_ = color;
    last_tick_ = std::chrono::steady_clock::now();
    running_ = true;
}

void Clock::stop() {
    if (running_) {
        consume_elapsed();
        running_ = false;
    }
}

void Clock::switch_side() {
    if (!active_color_.has_value()) {
        return;
    }
    if (running_) {
        consume_elapsed();
    }
    active_color_ = opposite(*active_color_);
    last_tick_ = std::chrono::steady_clock::now();
}

double Clock::remaining(Color color) const {
    const int idx = color_index(color);
    if (running_ && active_color_.has_value() && *active_color_ == color) {
        const auto now = std::chrono::steady_clock::now();
        const double elapsed =
            std::chrono::duration<double>(now - last_tick_).count();
        return std::max(0.0, remaining_[idx] - elapsed);
    }
    return std::max(0.0, remaining_[idx]);
}

bool Clock::is_flag_fallen(Color color) const {
    return remaining(color) <= 0.0;
}

void Clock::add_increment(Color color) {
    remaining_[color_index(color)] += time_control_.increment_seconds;
}

bool Clock::is_unlimited() const noexcept {
    return time_control_.is_unlimited();
}

void Clock::set_remaining(Color color, double seconds) {
    remaining_[color_index(color)] = seconds;
}

ClockSnapshot Clock::snapshot() const {
    return ClockSnapshot{
        remaining(Color::White),
        remaining(Color::Black),
        active_color_,
        running_,
    };
}

void Clock::restore(const ClockSnapshot& snapshot) {
    remaining_[color_index(Color::White)] = snapshot.white_remaining;
    remaining_[color_index(Color::Black)] = snapshot.black_remaining;
    active_color_ = snapshot.active_color;
    running_ = snapshot.is_running && snapshot.active_color.has_value();
    last_tick_ = std::chrono::steady_clock::now();
}

void Clock::consume_elapsed() {
    if (!active_color_.has_value()) {
        return;
    }
    const auto now = std::chrono::steady_clock::now();
    const double elapsed = std::chrono::duration<double>(now - last_tick_).count();
    const int idx = color_index(*active_color_);
    remaining_[idx] = std::max(0.0, remaining_[idx] - elapsed);
    last_tick_ = now;
}

}  // namespace chessie
